### PR TITLE
Pipeline

### PR DESCRIPTION
**Pull request by rssk**
# Pipeline
 ## How it works
Pipeline has four major components: manager, pipeline, controller, and computation.
### Manager
Controls pipelines themselves, allowing creating and in the future pausing/stoping/resuming. Manager also contains the code for decentralized network connections via a websocket interface.

### Pipeline
Takes a group of controllers and runs them end to end, feeding output into each successive step and also allowing mapping of inputs between steps.

### Controller
This can be thought of as a single Pipeline step, though controllers may either locally/decentralized iterate until a success condition is reached. The Controller step orchestrates the computation's runtime until that success condition, passing off control to the non-local node (whether the central node or the local/slave nodes) and waiting to resume local iterations.

### Computation
The manifestation of the computational part of a pipeline spec, this controls computation specific actions, notably at the moment this starting the docker box for the computation.
  
# Testing
`npm test`

### CODE REVIEW

File: packages/pipeline/computation.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+      meta,
+      mode,
+      start(input, { baseDirectory }) {
+        return docker.startService(
+          meta.id,
+          {
+            Image: computation.dockerImage,
+            HostConfig: {
+              Binds: [
+                `${baseDirectory}:/input:ro`,
```
[the remark this answers is not in the document]
Comment (rssk) [reply packages/pipeline/computation.js:25]: I think we talked about this, but the format for docker mounts is `hostDir:dockerDir:perms`

File: packages/pipeline/controller.js [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+              });
+            case 'nextComputation':
+              // TODO: code for multiple comps on one controller
+              // controllerState.computationIndex =
+              //   controllerState.computationIndex < currentComputations.length ?
+              //   controllerState.computationIndex + 1 : controllerState.computationIndex;
+              // run =
+              //   controllerState.activeComputations[controllerState.computationIndex]
+              //   .start(input);
+              break;
```
[the remark this answers is not in the document]
Comment (rssk) [reply packages/pipeline/controller.js:109]: Maaayyybbeee, I dunno yet, 50/50 on trashing

File: packages/pipeline/controller.js [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+                steps.push(lastCallback);
+              }
+              // necessary if this function call turns out synchronous
+              // the stack won't clear and overflow, has limited perf impact
+              process.nextTick(() => fn.apply(this, argsArray.concat(_cb)));
+              controllerState.remoteInitial = controllerState.mode === 'remote' ? false : undefined;
+            } :
+            undefined;  // steps complete
+          })(input);
+        };
```
[the remark this answers is not in the document]
Comment (rssk) [reply packages/pipeline/controller.js:200]: Definitely

File: packages/pipeline/pipeline-manager.js [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+        return memo;
+      }, {});
+    };
+
+    // TODO: secure socket layer
+    if (mode === 'remote') {
+      const app = http.createServer();
+      io = socketIO(app);
+
+      app.listen(3000);
```
[the remark this answers is not in the document]
Comment (rssk) [reply packages/pipeline/pipeline-manager.js]: 👍

File: packages/pipeline/pipeline-manager.js [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+      io.on('connection', (socket) => {
+        // TODO: not the way to do this, as runs would have to
+        // always start before clients connected....
+        // need proper auth
```
[the remark this answers is not in the document]
Comment (rssk) [reply packages/pipeline/pipeline-manager.js]: JWT from the auth server, commit for this incomming

**Comment by tkah** [comment]
Tests pass and coverage looks good! Excited to start integrating this with the rest of the application.

**Pull request merged**